Add deadline to createLiquidation()

mrice32 opened the pull request:
This adds a `deadline` parameter to `createLiquidation()` to protect callers from the price moving against them (and invalidating their liquidation) in the case that their transaction sits longer than expected in the mempool.

Note: the only logical changes to createLiquidation are the addition of the require at the top. Everything else is just restructuring to avoid the stack too deep error.

ptare commented:
Seems reasonable.

ptare commented:
> Looks like you're missing calls of `createLiquidation()` in `IntegrationTests.js`, `PrecisionErrors.js` and `ContractMonitors.js`.
> 
> I just did a search-all to find these

Yeah it look me hours yesterday when I added minTokens to hunt down all the tests.

By the way, do we still need minTokens as a parameter?

mrice32 commented:
> > Looks like you're missing calls of `createLiquidation()` in `IntegrationTests.js`, `PrecisionErrors.js` and `ContractMonitors.js`.
> > I just did a search-all to find these
> 
> Yeah it look me hours yesterday when I added minTokens to hunt down all the tests.
> 
> By the way, do we still need minTokens as a parameter?

I think some of these are in untested scripts, though :/. So 

> > Looks like you're missing calls of `createLiquidation()` in `IntegrationTests.js`, `PrecisionErrors.js` and `ContractMonitors.js`.
> > I just did a search-all to find these
> 
> Yeah it look me hours yesterday when I added minTokens to hunt down all the tests.
> 
> By the way, do we still need minTokens as a parameter?

Do you mean `minCollateralPerToken` or `maxTokensToLiquidate`?

ptare commented:
> > > Looks like you're missing calls of `createLiquidation()` in `IntegrationTests.js`, `PrecisionErrors.js` and `ContractMonitors.js`.
> > > I just did a search-all to find these
> > 
> > 
> > Yeah it look me hours yesterday when I added minTokens to hunt down all the tests.
> > By the way, do we still need minTokens as a parameter?
> 
> Do you mean `minCollateralPerToken` or `maxTokensToLiquidate`?

`minCollateralPerToken`, the one that we recently added. `max` makes sense to me, because deposits are instantaneous. But given we have a deadline, do we still need `min`? (I haven't thought deeply enough to have a strong opinion, but it seems both the min and the deadline solve similar problems).

mrice32 commented:
> Looks like you're missing calls of `createLiquidation()` in `IntegrationTests.js`, `PrecisionErrors.js` and `ContractMonitors.js`.
> 
> I just did a search-all to find these

Done.

mrice32 commented:
> > > > Looks like you're missing calls of `createLiquidation()` in `IntegrationTests.js`, `PrecisionErrors.js` and `ContractMonitors.js`.
> > > > I just did a search-all to find these
> > > 
> > > 
> > > Yeah it look me hours yesterday when I added minTokens to hunt down all the tests.
> > > By the way, do we still need minTokens as a parameter?
> > 
> > 
> > Do you mean `minCollateralPerToken` or `maxTokensToLiquidate`?
> 
> `minCollateralPerToken`, the one that we recently added. `max` makes sense to me, because deposits are instantaneous. But given we have a deadline, do we still need `min`? (I haven't thought deeply enough to have a strong opinion, but it seems both the min and the deadline solve similar problems).

They definitely have overlap IMO. I think the minCollateralPerToken solves both scenarios A and B (even though A is highly unlikely, as we've discussed). I think deadline just solves B. Like you, I don't have a strong opinion here.

mrice32 commented:
> > > > > Looks like you're missing calls of `createLiquidation()` in `IntegrationTests.js`, `PrecisionErrors.js` and `ContractMonitors.js`.
> > > > > I just did a search-all to find these
> > > > 
> > > > 
> > > > Yeah it look me hours yesterday when I added minTokens to hunt down all the tests.
> > > > By the way, do we still need minTokens as a parameter?
> > > 
> > > 
> > > Do you mean `minCollateralPerToken` or `maxTokensToLiquidate`?
> > 
> > 
> > `minCollateralPerToken`, the one that we recently added. `max` makes sense to me, because deposits are instantaneous. But given we have a deadline, do we still need `min`? (I haven't thought deeply enough to have a strong opinion, but it seems both the min and the deadline solve similar problems).
> 
> They definitely have overlap IMO. I think the minCollateralPerToken solves both scenarios A and B (even though A is highly unlikely, as we've discussed). I think deadline just solves B. Like you, I don't have a strong opinion here.

@ptare I'm going to go ahead and merge this, but maybe we can reconsider reverting the `minCollateralPerToken` change separately?

OliverNChalk commented:
Just weighing in on `minCollateralPerToken`, I am of the opinion the functionality should be removed. I agree that it seems like a much more fragile way of solving the issue `deadline` addresses

OliverNChalk commented:
By the way, my original recommendation was to add both `maxTimestamp` and `maxTimestamp`. The reason being that miners can fiddle with timestamps:

https://github.com/ethereum/wiki/blob/c02254611f218f43cbb07517ca8e5d00fd6d6d75/Block-Protocol-2.0.md

> Is `block.timestamp <= now + 900` and is `block.timestamp >= parent.timestamp`?

This means that if you send a liquidation as soon as the new 60-second candle prints, there is a non-negligible chance that you could be included in the previous candle and get penalized.

A further consequence of Ethereum's timestamp rules is that miners could break liquidations for 15 minutes if they set the block time to 900 seconds in the future at the same time a bunch of liquidations were about to happen.

Because there are two competing risks here (miner attack vs. miner negligence) and there are two ways of addressing these risks (have DVM side with the liquidator instead of the miner vs. implement minTimestamp measures), I will post my thoughts in slack for discussion.

mrice32 commented:
> By the way, my original recommendation was to add both `maxTimestamp` and `maxTimestamp`. The reason being that miners can fiddle with timestamps:
> 
> https://github.com/ethereum/wiki/blob/c02254611f218f43cbb07517ca8e5d00fd6d6d75/Block-Protocol-2.0.md
> 
> > Is `block.timestamp <= now + 900` and is `block.timestamp >= parent.timestamp`?
> 
> This means that if you send a liquidation as soon as the new 60-second candle prints, there is a non-negligible chance that you could be included in the previous candle and get penalized.
> 
> A further consequence of Ethereum's timestamp rules is that miners could break liquidations for 15 minutes if they set the block time to 900 seconds in the future at the same time a bunch of liquidations were about to happen.
> 
> Because there are two competing risks here (miner attack vs. miner negligence) and there are two ways of addressing these risks (have DVM side with the liquidator instead of the miner vs. implement minTimestamp measures), I will post my thoughts in slack for discussion.

Thanks for the feedback! I picked up the conversation in slack.
